loadtest/loadtest.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package loadtest
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/mattermost/mattermost-load-test-ng/config"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/control"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+// LoadTester is a structure holding all the state needed to run a load-test
+type LoadTester struct {
+	controllers   []control.UserController
+	config        *config.LoadTestConfig
+	wg            sync.WaitGroup
+	status        chan control.UserStatus
+	started       bool
+	newController NewController
+}
+
+// NewController is a factory function that returns a new
+// control.UserController given an id and a channel of control.UserStatus
+type NewController func(int, chan<- control.UserStatus) control.UserController
+
+func (lt *LoadTester) handleStatus() {
+	for us := range lt.status {
+		if us.Code == control.USER_STATUS_STOPPED || us.Code == control.USER_STATUS_FAILED {
+			lt.wg.Done()
+		}
+		if us.Code == control.USER_STATUS_ERROR {
+			mlog.Info(us.Err.Error(), mlog.Int("controller_id", us.ControllerId))
+			continue
+		} else if us.Code == control.USER_STATUS_FAILED {
+			mlog.Error(us.Err.Error())
+			continue
+		}
+		mlog.Info(us.Info, mlog.Int("controller_id", us.ControllerId))
+	}
+}
+
+// AddUser increments by one the number of concurrently active users
+func (lt *LoadTester) AddUser() error {
+	if !lt.started {
+		return errors.New("LoadTester is not running")
+	}
+	activeUsers := len(lt.controllers)
+	if activeUsers == lt.config.UsersConfiguration.MaxActiveUsers {
+		return errors.New("Max active users limit reached")
+	}
+	controller := lt.newController(activeUsers+1, lt.status)
+	lt.wg.Add(1)
+	go func() {
+		controller.Run()
+	}()
+	lt.controllers = append(lt.controllers, controller)
```

Comment:
There is a possible race condition here during AddUser/RemoveUser if they get called from different goroutines. It may be worthwhile to protect the access to `controllers` slice with a mutex.

Reply:
Yes, this is another case where I've intentionally left some decisions for later.  
Access to `LoadTester.started` should also be atomic/guarded if we ever wanted to use it from different goroutines.  
I'd like to review this when we have a better picture of how we are going to use this API to avoid overengineering it now and wait until we have a concrete reason to do so.

Reply:
That's fine. I would suggest to just leave a comment noting down this thought, so that it is apparent for someone reading the code.